azure-kusto-data/azure/kusto/data/helpers.py

```diff
@@ -48,33 +80,25 @@ def dataframe_from_result_table(table: "Union[KustoResultTable, KustoStreamingRe
 
     columns = [col.column_name for col in table.columns]
     frame = pd.DataFrame(table.raw_rows, columns=columns)
+    default = default_dict()
+
+    if nullable_bools:
+        default = copy.deepcopy(default)
+        default["bool"] = lambda col, df: df[col].astype(pd.BooleanDtype())
 
-    # fix types
     for col in table.columns:
-        if col.column_type == "string" and hasattr(pd, "StringDType"):
-            frame[col.column_name] = frame[col.column_name].astype(pd.StringDType())
-        if col.column_type == "bool":
-            frame[col.column_name] = frame[col.column_name].astype(pd.BooleanDtype() if nullable_bools else bool)
-        elif col.column_type == "int":
-            frame[col.column_name] = frame[col.column_name].astype(pd.Int32Dtype())
-        elif col.column_type == "long":
-            frame[col.column_name] = frame[col.column_name].astype(pd.Int64Dtype())
-        elif col.column_type == "real" or col.column_type == "decimal":
-            frame[col.column_name] = frame[col.column_name].replace("NaN", np.NaN).replace("Infinity", np.PINF).replace("-Infinity", np.NINF)
-            frame[col.column_name] = pd.to_numeric(frame[col.column_name], errors="coerce").astype(pd.Float64Dtype())
-        elif col.column_type == "datetime":
-            # Pandas before version 2 doesn't support the "format" arg
-            args = {}
-            if pd.__version__.startswith("2."):
-                args = {"format": "ISO8601", "utc": True}
+        if converters_by_column_name and col.column_name in converters_by_column_name.keys():
+            if isinstance(converters_by_column_name[col.column_name], str):
+                frame[col.column_name] = frame[col.column_name].astype(converters_by_column_name[col.column_name])
             else:
-                # if frame contains ".", replace "Z" with ".000Z"
-                # == False is not a mistake - that's the pandas way to do it
-                contains_dot = frame[col.column_name].str.contains(".")
-                frame.loc[contains_dot == False, col.column_name] = frame.loc[contains_dot == False, col.column_name].str.replace("Z", ".000Z")
-            frame[col.column_name] = pd.to_datetime(frame[col.column_name], errors="coerce", **args)
-        elif col.column_type == "timespan":
-            frame[col.column_name] = frame[col.column_name].apply(to_pandas_timedelta)
+                frame[col.column_name] = converters_by_column_name[col.column_name](col.column_name, frame)
+        elif converters_by_type and col.column_type in converters_by_type.keys():
+            if isinstance(converters_by_type[col.column_type], str):
+                frame[col.column_name] = frame[col.column_name].astype(converters_by_type[col.column_type])
+            else:
+                frame[col.column_name] = converters_by_type[col.column_type](col.column_name, frame)
+        else:
+            frame[col.column_name] = default[col.column_type](col.column_name, frame)
 
```

Comment:
This code is a bit repetitive. You can have a much nicer loop if you separate the retrival with the usage:


```suggestion
    for col in table.columns:
        column_name = col.column_name
        column_type = col.column_type

        if converters_by_column_name and column_name in converters_by_column_name:
            converter = converters_by_column_name[column_name]
        elif converters_by_type and column_type in converters_by_type:
            converter = converters_by_type[column_type]
        elif nullable_bools and column_type == "bool":
                converter = lambda col, df: df[col].astype(pd.BooleanDtype())
        else:
            converter = default[column_type]

        if isinstance(converter, str):
            frame[column_name] = frame[column_name].astype(converter)
        else:
            frame[column_name] = converter(column_name, frame)
```